components/gif_picker/components/InfiniteScroll/index.tsx

```diff
@@ -0,0 +1,187 @@
+// See LICENSE.txt for license information.
+
+import React, {PureComponent} from 'react';
+
+type BaseElementProps = {
+    className?: string;
+}
+
+type InfiniteScrollDefaultProps = {
+    element: keyof JSX.IntrinsicElements;
+    hasMore: boolean;
+    initialLoad: boolean;
+    pageStart: number;
+    threshold: number;
+    useWindow: boolean;
+    isReverse: boolean;
+    containerHeight: number | null;
+    scrollPosition: number | null;
+}
+
+type InfiniteScrollProps = BaseElementProps & InfiniteScrollDefaultProps & {
+    children: React.ReactNode;
+    loader?: React.ReactNode;
+    loadMore: (page: number) => void;
+}
+
+type InfiniteScrollElementProps = BaseElementProps & {
+    ref: (node: HTMLElement) => void;
+}
+
+const isHtmlElement = (data: HTMLElement | Element | null | undefined): data is HTMLElement => {
+    return data != null;
+};
+
+export default class InfiniteScroll extends PureComponent<InfiniteScrollProps> {
+    scrollComponent: HTMLElement | undefined;
+    defaultLoader: React.ReactNode = null;
+    pageLoaded = 0;
+
+    static defaultProps: InfiniteScrollDefaultProps = {
+        element: 'div' as keyof JSX.IntrinsicElements,
+        hasMore: false,
+        initialLoad: true,
+        pageStart: 0,
+        threshold: 250,
+        useWindow: true,
+        isReverse: false,
+        containerHeight: null,
+        scrollPosition: null,
+    }
+
+    componentDidMount() {
+        this.pageLoaded = this.props.pageStart ?? 0;
+        this.attachScrollListener();
+        this.setScrollPosition();
+    }
+
+    componentDidUpdate() {
+        this.attachScrollListener();
+    }
+
+    render() {
+        const {
+            children,
+            element,
+            hasMore,
+            initialLoad, // eslint-disable-line @typescript-eslint/no-unused-vars
+            loader,
+            loadMore, // eslint-disable-line @typescript-eslint/no-unused-vars
+            pageStart, // eslint-disable-line @typescript-eslint/no-unused-vars
+            threshold, // eslint-disable-line @typescript-eslint/no-unused-vars
+            useWindow, // eslint-disable-line @typescript-eslint/no-unused-vars
+            isReverse, // eslint-disable-line @typescript-eslint/no-unused-vars
+            scrollPosition, // eslint-disable-line @typescript-eslint/no-unused-vars
+            containerHeight,
+            ...rest
+        } = this.props;
+
+        const baseElementProps: InfiniteScrollElementProps = {
+            ref: (node) => {
+                this.scrollComponent = node;
+            },
+            ...rest,
+        };
+
+        const elementProps = containerHeight ? {...baseElementProps, style: {height: containerHeight}} : baseElementProps;
+
+        return React.createElement(element, elementProps, children, hasMore && (loader || this.defaultLoader));
+    }
+
+    calculateTopPosition(el: HTMLElement | Element | null | undefined): number {
+        if (!isHtmlElement(el)) {
```

Comment:
`offsetTop` and `offsetParent` only exists in HTMLElement type  but `offsetParent` returns an Element. As this is a recursive function, I'm using a type guard to narrow the type. This case we know at runtime is truthy but at compile time not.